### PR TITLE
[FIX] mrp: improve perf of _compute_quantities and explode

### PR DESCRIPTION
**Pull request by Aurelienvd**
Batch the bom_search of bom_lines in mrp_bom.explode.
If we represent the bom hierarchy as a graph, essentially
switch from depth-first to breadth-first, allowing
to batch search boms for each depth level, hence improving
the overall performance.

The part around psql_sequence is added because according to mrp_bom explode tests, the sequence field of mrp_bom defines
which bom should be returned during explode.  It is only in case of ties that the bom with product_id takes precedence.

##### Speedup

In client DB with 15k products and 2k boms, negative forecasted quantity filter on product variants goes
from `50s -> ~12s`.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/product.py [hunk 1]
```diff
@@ -109,7 +109,12 @@ def _compute_quantities_dict(self, lot_id, owner_id, package_id, from_date=False
         This override is used to get the correct quantities of products
         with 'phantom' as BoM type.
         """
-        bom_kits = self.env['mrp.bom']._get_product2bom(self, bom_type='phantom')
+        products = self.filtered(lambda p: p.type  != 'service')
+        bom_kits = {
+            product: bom
+            for product, bom in zip(products, self.env['mrp.bom']._get_product2bom(products, bom_type='phantom'))
+            if bom
+        }
```
Comment (yelizariev) [inline addons/mrp/models/product.py:117]: isn't the same as 

        products = self.filtered(lambda p: p.type  != 'service')
        bom_kits = self.env['mrp.bom']._get_product2bom(products, bom_type='phantom')


?￼

**Review by Elkasitu — CHANGES_REQUESTED** [review]
requires some modifications and clarifications

File: addons/mrp/models/product.py (hunk 1 above)
Comment (Elkasitu) [reply addons/mrp/models/product.py:117]: the previous version of `_get_product2bom` yes, the version introduced by this patch returns either a list of boms, list of `False` or empty recordset which while it's one of Python's strengths, is poor in practice (I'd rather have a single return type if possible)

File: addons/mrp/models/mrp_bom.py [hunk 2]
```diff
@@ -192,34 +192,44 @@ def _bom_find(self, product_tmpl=None, product=None, picking_type=None, company_
         return self.search(domain, order='sequence, product_id', limit=1)
 
     @api.model
-    def _get_product2bom(self, products, bom_type=False):
+    def _get_product2bom(self, products, picking_type=False, company_id=False, bom_type=False):
```
[the remark this answers is not in the document]
Comment (Aurelienvd) [reply addons/mrp/models/mrp_bom.py]: I removed the `company_id` param as it wasn't useful anyway and I moved `picking_type` param to the end. However I don't really see how I can avoid adding this parameter 😕 This function was added by Ivan last week to improve the computation of the product quantities. In `mrp/product._computate_quantities_dict` there was no need for the `picking_type` but this is not the case for `mrp.explode`.

I don't see how I could avoid it by filtering boms after calling `_get_product2bom` as it only returns 1 bom per product.

File: addons/mrp/models/mrp_bom.py [hunk 3]
```diff
@@ -192,34 +192,44 @@ def _bom_find(self, product_tmpl=None, product=None, picking_type=None, company_
...
         for p in products:
-            bom = variant2bom.get(p) or template2bom.get(p.product_tmpl_id)
-            if bom:
-                result[p] = bom
+            if p.type == 'service':
+                result.append(False)
+            elif template2bom.get(p.product_tmpl_id, {'sequence': float('inf')})['sequence'] < variant2bom.get(p, {'sequence':float('inf')})['sequence']:
+                result.append(template2bom.get(p.product_tmpl_id, {}).get('bom'))
+            else:
+                result.append(variant2bom.get(p, {}).get('bom') or template2bom.get(p.product_tmpl_id, {}).get('bom'))
```
[the remark this answers is not in the document]
Comment (Aurelienvd) [reply addons/mrp/models/mrp_bom.py]: That's because `_get_product2bom` is basically a batch version of `_bom_find`. `_bom_find` is supposed to return 1 bom for a given product, with the `sequence` param being the first discriminant.

If we have the following data

| bom_id | sequence | product_tmpl_id | product_id |
|:-------:|:----------:|:-----------------:|:-----------:|
| 1|0|1| |
| 2|1|1|1|

Since `mrp.bom(1)` has a lower sequence, `_bom_find` will return `mrp.bom(1)` for product.product(1). Currently, `_get_product2bom` will return `mrp.bom(2)` because it prioritizes boms with `product_id != False` while not accounting for `sequence` values.

This did not cause any issue when merging `_get_product2bom` because `product._computate_quantities_dict` didn't use that. But there are some unit tests for `mrp.bom.explode` that failed because of that.

**Comment by yelizariev** [comment]
opw-2449630

File: addons/mrp/models/mrp_bom.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Elkasitu) [reply addons/mrp/models/mrp_bom.py]: Don't worry about it, it was just a note. With the current implementation the method signature change is as safe as can be for a stable branch

File: addons/mrp/models/mrp_bom.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Elkasitu) [reply addons/mrp/models/mrp_bom.py]: I see, thanks for the clarification

**Comment by ryv-odoo** [comment]
In saas-14.3 and master, the `_bom_find` is now batch (then _get_product2bom doesn't exist). the explode will be refactor soon in master with the new version of `_bom_find` (for performance and clean code).
@fw-bot up to saas-14.2

File: addons/mrp/models/mrp_bom.py [hunk 4]
```diff
@@ -207,19 +211,25 @@ def _get_product2bom(self, products, bom_type=False):
         boms = self.search(domain, order='sequence, product_id')
         template2bom = {}
         variant2bom = {}
+        # Cannot use bom.sequence as unset sequence values are set to 0.
+        # Use counter to maintain the order of self.search above.
+        psql_sequence = 0
```
Comment (ryv-odoo) [inline addons/mrp/models/mrp_bom.py:216]: It should work but
IMO, it is become a complicate (a bit messy) code. Maybe back-port the batch version of `_bom_find` (https://github.com/odoo/odoo/commit/974c4398b4eb3bb9fa097632e5223bd9efb6811c) in this method from master which it covers the case of order in account. What do you think?

**Review by ryv-odoo — COMMENTED** [review]
The change in explode seems legit and a nice improvements in term of performance.
But the `_get_product2bom` become complicate even it seems correct.

File: addons/mrp/models/mrp_bom.py (hunk 4 above)
Comment (Aurelienvd) [reply addons/mrp/models/mrp_bom.py]: Yeah sure, it does look neater. I didn't know there was a new batch version of _bom_find starting 14.3 😮

File: addons/mrp/models/mrp_bom.py [hunk 5]
```diff
@@ -192,35 +193,37 @@ def _bom_find(self, product_tmpl=None, product=None, picking_type=None, company_
...
-            bom = variant2bom.get(p) or template2bom.get(p.product_tmpl_id)
+        if len(products) == 1:
+            bom = self.search(domain, order='sequence, product_id, id', limit=1)
             if bom:
-                result[p] = bom
-        return result
+                bom_by_product[products] = bom
+            return bom_by_product
+
+        boms = self.search(domain, order='sequence, product_id, id')
```
Comment (ryv-odoo) [inline addons/mrp/models/mrp_bom.py:218]: Don't change the order in stable (not consistent with the `_bom_find()` )

File: addons/mrp/models/product.py [hunk 6]
```diff
@@ -109,7 +109,8 @@ def _compute_quantities_dict(self, lot_id, owner_id, package_id, from_date=False
         This override is used to get the correct quantities of products
         with 'phantom' as BoM type.
         """
-        bom_kits = self.env['mrp.bom']._get_product2bom(self, bom_type='phantom')
+        products = self.filtered(lambda p: p.type  != 'service')
+        bom_kits = {k:v for (k,v) in self.env['mrp.bom']._get_product2bom(products, bom_type='phantom').items() if v}
```
Comment (ryv-odoo) [inline addons/mrp/models/product.py:113]: you can simplify by `bom_kits = self.env['mrp.bom']._get_product2bom(products, bom_type='phantom')` because of the defaultdict return by the `_get_product2bom`

File: addons/mrp/models/mrp_bom.py [hunk 7]
```diff
@@ -261,15 +270,20 @@ def check_cycle(v, visited, recStack, graph):
...
             line_quantity = current_qty * current_line.product_qty
-            bom = self._bom_find(product=current_line.product_id, picking_type=picking_type or self.picking_type_id, company_id=self.company_id.id, bom_type='phantom')
+            if not current_line.product_id in product_boms:
+                product_boms.update(self._get_product2bom(products_to_search, bom_type='phantom', picking_type=picking_type or self.picking_type_id))
+                products_to_search = self.env['product.product']
+            bom = product_boms.get(current_line.product_id)
             if bom:
                 converted_line_quantity = current_line.product_uom_id._compute_quantity(line_quantity / bom.product_qty, bom.product_uom_id)
-                bom_lines = [(line, current_line.product_id, converted_line_quantity, current_line) for line in bom.bom_line_ids] + bom_lines
+                bom_lines = bom_lines + [(line, current_line.product_id, converted_line_quantity, current_line) for line in bom.bom_line_ids]
```
Comment (ryv-odoo) [inline addons/mrp/models/mrp_bom.py:279]: Why this change ? (if it is legit, maybe use `+=`)

File: addons/mrp/models/mrp_bom.py [hunk 8]
```diff
@@ -249,10 +252,16 @@ def check_cycle(v, visited, recStack, graph):
...
+        products_to_search = self.env['product.product']
         for bom_line in self.bom_line_ids:
-            V |= set([bom_line.product_id.product_tmpl_id.id])
-            graph[product.product_tmpl_id.id].append(bom_line.product_id.product_tmpl_id.id)
+            product_id = bom_line.product_id
+            V |= set([product_id.product_tmpl_id.id])
+            graph[product.product_tmpl_id.id].append(product_id.product_tmpl_id.id)
+            bom_lines.append((bom_line, product, quantity, False))
+            products_to_search |= product_id
+        product_boms = self._get_product2bom(products_to_search, bom_type='phantom', picking_type=picking_type or self.picking_type_id)
```
Comment (ryv-odoo) [inline addons/mrp/models/mrp_bom.py:263]: You remove the `company_id` argument here. Then `_get_product2bom` will take the `company_id` in context (if there is) before the `company_id` argument (here == to `self.company_id.id`). I am not sure if the `company_id` from context is use somewhere but it is still a behavior change of the method. Or I miss something ?

File: addons/mrp/models/mrp_bom.py [hunk 9]
```diff
@@ -249,10 +252,16 @@ def check_cycle(v, visited, recStack, graph):
...
+        bom_lines = []
+        products_to_search = self.env['product.product']
         for bom_line in self.bom_line_ids:
-            V |= set([bom_line.product_id.product_tmpl_id.id])
-            graph[product.product_tmpl_id.id].append(bom_line.product_id.product_tmpl_id.id)
+            product_id = bom_line.product_id
+            V |= set([product_id.product_tmpl_id.id])
+            graph[product.product_tmpl_id.id].append(product_id.product_tmpl_id.id)
+            bom_lines.append((bom_line, product, quantity, False))
+            products_to_search |= product_id
```
Comment (ryv-odoo) [inline addons/mrp/models/mrp_bom.py:262]: I am concerned about the union of the recordset due to his complexity (N²) - (I can be issue if there is more than 5K lines, it is very unlikely, but we are always surprised by the client data quantity). Maybe a OrderedSet of the ids is maybe safer?

File: addons/mrp/models/product.py [hunk 10]
```diff
@@ -109,7 +109,8 @@ def _compute_quantities_dict(self, lot_id, owner_id, package_id, from_date=False
         This override is used to get the correct quantities of products
         with 'phantom' as BoM type.
         """
-        bom_kits = self.env['mrp.bom']._get_product2bom(self, bom_type='phantom')
+        products = self.filtered(lambda p: p.type  != 'service')
```
Comment (ryv-odoo) [inline addons/mrp/models/product.py:112]: Not need to filter out here. It is already the case in `_get_product2bom`

**Review by ryv-odoo — COMMENTED** [review]
Few comments.

File: addons/mrp/models/mrp_bom.py (hunk 8 above)
Comment (Aurelienvd) [reply addons/mrp/models/mrp_bom.py]: No no, I indeed forgot the company_id param.

File: addons/mrp/models/mrp_bom.py (hunk 9 above)
Comment (Aurelienvd) [reply addons/mrp/models/mrp_bom.py]: An id set is better but do we really need an OrderedSet here ? I know that recordsets keep order but here the purpose of products_to_search is to save encountered product_ids for _get_product2bom. Since this function returns an unordered dict wouldn't a simple `set` be enough ?

**Comment by Aurelienvd** [comment]
@ryv-odoo Just to inform you, I'm still checking things for this PR so it shouldn't be r+ yet

File: addons/mrp/models/mrp_bom.py (hunk 9 above)
Comment (ryv-odoo) [reply addons/mrp/models/mrp_bom.py]: Yes you are perfectly right :+1: , we don't need to conserve the order.

File: addons/mrp/models/mrp_bom.py (hunk 7 above)
Comment (Aurelienvd) [reply addons/mrp/models/mrp_bom.py]: This allows to group the lines product_id of the same hierarchy level together, reducing the number of calls to _get_product2bom. If we have 1 bom with 4 bom_lines, each one with a bom with 2 bom_lines, earch one with a bom, kinda like this:
<pre>
.bom 1
+--- bom 2
|      +--- bom 6
|      +--- bom 7
+--- bom 3
|      +--- bom 8
|      +--- bom 9
+--- bom 4
|      +--- bom 10
|      +--- bom 11
+--- bom 5
|      +--- bom 12
|      +--- bom 13
</pre>

The last level will be grouped together, reducing the number of calls. In this example, the number of calls is theoretically equal to 3 (the height of the tree+1) while it would be 13 without the change (1 call for the first level after root and 1 for each sibling group in the tree). I say theoretically because with the backport of 14.3 there are more calls to product2boms but most of them enters the first if condition.

In practice I don't think that it will change the perf that much because most of the time there are only one or two levels in the bom hierarchy but it is still worth it I think.

PS: I changed the line to `bom_lines += ...`

**Comment by Aurelienvd** [comment]
@ryv-odoo I added a small nested function to explode to add missing product keys to `product_boms`. This is because `product_boms` is supposed to store a bom for each encountered product. Adding missing product keys avoids calling `get_product2bom` multiple times if the same product without bom is used in multiple boms at different hierarchy levels.

I could have added it to `get_product2bom` but I thought it was better to avoid bloating this method.

File: addons/mrp/models/mrp_bom.py [hunk 11]
```diff
@@ -245,14 +248,29 @@ def check_cycle(v, visited, recStack, graph):
...
         for bom_line in self.bom_line_ids:
-            V |= set([bom_line.product_id.product_tmpl_id.id])
-            graph[product.product_tmpl_id.id].append(bom_line.product_id.product_tmpl_id.id)
+            product_id = bom_line.product_id
+            V |= set([product_id.product_tmpl_id.id])
+            graph[product.product_tmpl_id.id].append(product_id.product_tmpl_id.id)
+            bom_lines.append((bom_line, product, quantity, False))
+            product_ids.add(product_id.id)
+        update_product_boms(product_boms, product_ids)
+        product_ids = set()
```
Comment (ryv-odoo) [inline addons/mrp/models/mrp_bom.py:273]: Maybe use product_ids.clear() to be more readable?

File: addons/mrp/models/mrp_bom.py [hunk 12]
```diff
@@ -245,14 +248,29 @@ def check_cycle(v, visited, recStack, graph):
             recStack[v] = False
             return False
 
+        def update_product_boms(product_boms: defaultdict, product_ids):
+            products = self.env['product.product'].browse(product_ids)
+            product_boms.update(self._get_product2bom(products, bom_type='phantom',
+                picking_type=picking_type or self.picking_type_id, company_id=self.company_id.id))
+            # Set missing keys to default value
```
Comment (ryv-odoo) [inline addons/mrp/models/mrp_bom.py:255]: it is weird to use a defaultdict if it is to force the missing key for each product, maybe use a simple dict. Also if you want to keep your inner method, instantiate the product_boms before to avoid pass it in arguments (also for product_ids ???).
```python
product_boms = {}

def update_product_boms(product_ids):
    products = self.env['product.product'].browse(product_ids)
    product_boms.update(dict.fromkeys(products, False))
    product_boms.update(self._get_product2bom(products, bom_type='phantom', picking_type=picking_type or self.picking_type_id, company_id=self.company_id.id))
```

**Review by ryv-odoo — COMMENTED** [review]
Thanks for your explanation about the concat of `bom_lines` (nice, make sense).
I let you 2 comments about the clarity of the code.

File: addons/mrp/models/mrp_bom.py (hunk 12 above)
Comment (Aurelienvd) [reply addons/mrp/models/mrp_bom.py:257]: Yeah sure, I also did it for product_ids so that `update_product_boms` does not take any argument anymore

**Comment by Aurelienvd** [comment]
@ryv-odoo I pushed the changes requested in the last review. I think it is ready to be merged, at least I'm done testing in client databases. Are there still some changes you think should be made to the code ?

**Comment by yelizariev** [comment]
@Aurelienvd have you tried with other customer databases? 

Maybe we still need further performance improvements like I suggest in https://github.com/odoo-dev/odoo/pull/728 ?
image

**Comment by Aurelienvd** [comment]
@yelizariev  Yeah I tested it with ticket 2449633, 2502719, 2449630 and 2522267 databases.

Regarding odoo-dev#728 , I tried to make it work locally but I didn't manage to attain a huge performance boost for the customer databases that I tested. Maybe I missed something but batching the explode is tricky because the `picking_type` and `company_id` are both bom-dependent.

So I thought that it would be useful to deploy this PR first since there are multiple customers having issues with `explode` and maybe try optimizing it further afterwards if possible.

I can try batching each level of the bom hierarchy "forest" level instead of each tree level to make it work in batch but that would require to complexify the `get_product2bom` function. But if you have another suggestion on how to do it I can also give it a try.

**Comment by yelizariev** [comment]
ok, let's stay with current version

**Review by ryv-odoo — APPROVED** [review]

**Review by ryv-odoo — COMMENTED** [review]
To avoid to miss this optimisation for master (I don't know when the refactor will be done), I reset the forward bot to master. But for the version saas143.3 and master `_get_product2bom` should be change by the `_bom_find`.
@fw-bot up to master

**Comment by Whenrow** [comment]
Thanks everyone
robodoo r+
